linkerd/app/inbound/src/lib.rs

```diff
@@ -290,11 +292,16 @@ impl<A: OrigDstAddr> Config<A> {
                 http_server.into_inner(),
                 h2_settings,
                 drain.clone(),
-                disable_protocol_detection_for_ports.clone(),
             );
 
+            let svc = svc::stack(tcp_server)
+                .push(DetectProtocolLayer::new(ProtocolDetect::new(
+                    disable_protocol_detection_for_ports.clone(),
+                )))
+                .push_timeout(detect_protocol_timeout);
```

Comment:
I just realized this is unlikely to work as expected... The accept stack handles processing of the request, so this timeout seems like it will limit the _lifetime_ of all TCP connections...

I think we have two options:

1. Handle the timeout in the detect layer. The advantage to this approach would be that we can return the underlying socket in the error so that it could be rerouted if we want to do that later.
2. If we want to keep the timeout separate; we'd have to model Detect to have a response type that is the detected protocol and the socket; and then we'd have another layer that basically does `detect_or_timeout.call(sock).and_then(self.inner.accept)`.

I think, all things considered, it might be simplest to add the timeout to the detection layer directly.